### PR TITLE
Docker GitHub actions

### PR DESCRIPTION
**Pull request by hennis1**
# Summary

- This PR is for adding github actions to build axom in gcc8 and clang10 dockerfile images.
- It does the following (modify list as needed):
  - Modifies docker files.
  - Adds github action yml file at the request of Chris White.

# Design review (for API changes or additions---delete if unneeded)

On (date), we reviewed this PR.  We discussed the design ideas:

1. Add github actions from serac to axom.
2. Get the builds working with adjustments to dockerfiles and build files.
3. Get the hostconfig file.

This PR implements 1.  It leaves out 2 & 3. for the following reasons

- WIP

### CODE REVIEW

**Comment by white238** [comment]
Clang: 

Currently looks like there is an hdf5 configuration/build error

```
2020-12-01T01:42:02.2593088Z #13 355.2 [ RUN      ] spio_serial.basic_writeread
2020-12-01T01:42:02.2593832Z #13 355.2 HDF5-DIAG: Error detected in HDF5 (1.8.21) MPI-process 0:
2020-12-01T01:42:02.2594767Z #13 355.2   #000: H5E.c line 1591 in H5Eget_auto2(): library initialization failed
2020-12-01T01:42:02.2595344Z #13 355.2     major: Function entry/exit
2020-12-01T01:42:02.2595839Z #13 355.2     minor: Unable to initialize object
2020-12-01T01:42:02.2596446Z #13 355.2   #001: H5.c line 210 in H5_init_library(): unable to initialize dataset interface
2020-12-01T01:42:02.2597036Z #13 355.2     major: Function entry/exit
2020-12-01T01:42:02.2597509Z #13 355.2     minor: Unable to initialize object
2020-12-01T01:42:02.2598114Z #13 355.2   #002: H5Dint.c line 141 in H5D_init(): interface initialization failed
2020-12-01T01:42:02.2598677Z #13 355.2     major: Function entry/exit
2020-12-01T01:42:02.2599162Z #13 355.2     minor: Unable to initialize object
2020-12-01T01:42:02.2600181Z #13 355.2   #003: H5Dint.c line 183 in H5D__init_interface(): can't get default dataset creation property list
2020-12-01T01:42:02.2600771Z #13 355.2     major: Dataset
2020-12-01T01:42:02.2601211Z #13 355.2     minor: Inappropriate type
2020-12-01T01:42:02.2601586Z #13 355.2 
2020-12-01T01:42:02.2601891Z #13 355.2 ***********************************
2020-12-01T01:42:02.2603522Z #13 355.2 [ERROR in line 2490 of file /home/axom/axom_tpls/builds/spack-stage-conduit-master-5fomtqyji7khjpzwirhwaobqy5uujn4i/spack-src/src/libs/relay/conduit_relay_io_hdf5.cpp]
2020-12-01T01:42:02.2605292Z #13 355.2 MESSAGE=HDF5 Error code-1 Failed to create H5P_FILE_CREATE  property list
2020-12-01T01:42:02.2605899Z #13 355.2 ** StackTrace of 24 frames **
```

**Comment by white238** [comment]
gcc looks like its passing but failing some permission.. maybe.

```
2020-12-01T03:02:41.3638563Z #21 pushing layers
2020-12-01T03:02:41.7465550Z #21 pushing layers 0.4s done
2020-12-01T03:02:41.7467250Z #21 ERROR: server message: insufficient_scope: authorization failed
```

File: .github/workflows/docker_build_tpls.yml [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  workflow_dispatch:
+    # No input required
+
+jobs:
+  clang-10:
+    runs-on: ubuntu-latest
+    name: Builds a Clang 10 docker image and extracts generated hostconfigs
+    env:
+      TAG: clang-10
+      REPO: axomllnl/tpls
```
Comment (joshessman-llnl) [inline .github/workflows/docker_build_tpls.yml:14]: I think this should be just `axom/tpls` - https://hub.docker.com/r/axom/tpls

Comment (hennis1) [reply .github/workflows/docker_build_tpls.yml]: Nice that fixed it! gcc-8 now works. now to fix clang...

Comment (white238) [reply .github/workflows/docker_build_tpls.yml]: its not liking hdf5 for some reason....

Comment (joshessman-llnl) [reply .github/workflows/docker_build_tpls.yml]: Serac had a very similar issue: https://dev.azure.com/llnl-serac/serac/_build/results?buildId=2166&view=logs&j=2742912e-f547-59b4-f336-65c28301f30c&t=cd7bfce2-9b19-5c4f-5f3a-913e42776d4e&l=1341

```
HDF5-DIAG: Error detected in HDF5 (1.8.21) MPI-process 0:
  #000: H5E.c line 1653 in H5Eset_auto2(): library initialization failed
    major: Function entry/exit
    minor: Unable to initialize object
  #001: H5.c line 210 in H5_init_library(): unable to initialize dataset interface
    major: Function entry/exit
    minor: Unable to initialize object
  #002: H5Dint.c line 141 in H5D_init(): interface initialization failed
    major: Function entry/exit
    minor: Unable to initialize object
  #003: H5Dint.c line 183 in H5D__init_interface(): can't get default dataset creation property list
    major: Dataset
    minor: Inappropriate type
HDF5-DIAG: Error detected in HDF5 (1.8.21) MPI-process 0:
  #000: H5P.c line 297 in H5Pcreate(): not a property list class
    major: Invalid arguments to routine
    minor: Inappropriate type


MFEM abort: Fatal NetCDF error: NetCDF: HDF error
 ... in function: void mfem::Mesh::ReadCubit(const char *, int &, int &)
 ... in file: mesh/mesh_readers.cpp:2187

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
```

Comment (joshessman-llnl) [reply .github/workflows/docker_build_tpls.yml]: Looking at this spec:
```
axom@develop%clang@10.0.0+cpp14~cuda~debug~devtools+fortran+hdf5+lua~mfem+mpi+openmp~python+raja~scr+shared+umpire build_type=RelWithDebInfo cuda_arch=none arch=linux-ubuntu18.04-x86_64
  ^conduit@master%clang@10.0.0~adios~doc~doxygen+fortran+hdf5+hdf5_compat+mpi~python+shared~silo+test~zfp arch=linux-ubuntu18.04-x86_64
     ^hdf5@1.8.21%clang@10.0.0~cxx~debug~fortran~hl~java+mpi+pic~shared~szip~threadsafe api=none patches=b61e2f058964ad85be6ee5ecea10080bf79e73f83ff88d1fa4b602d00209da9c arch=linux-ubuntu18.04-x86_64
```
I wonder if changing HDF5 to be a shared library might help (s.t. they're all the same - changing them to all static might be interesting too)

Comment (cyrush) [reply .github/workflows/docker_build_tpls.yml]: this seems like an issue with multiple versions of hdf5 (or multiple links of hdf5 at play).

I think a mix of shared and static could be the source of woes.

File: scripts/docker/dockerfile_gcc-8 [hunk 2]
```diff
@@ -1,25 +1,38 @@
 FROM axom/compilers:gcc-8
+LABEL maintainer="white238@llnl.gov"
+ARG branch=develop
 
+SHELL ["/bin/bash", "-c"]
 RUN sudo apt-get update -y
-RUN sudo apt-get install curl -fy
-
+RUN sudo apt-get install curl wget ssh pkg-config build-essential libopenblas-dev tar python3 -fy
 RUN sudo apt-get -qq install -y --no-install-recommends gfortran-7 && sudo update-alternatives --install /usr/bin/gfortran gfortran /usr/bin/gfortran-7 100
```
Comment (kennyweiss) [inline scripts/docker/dockerfile_gcc-8:8]: Probably outside the scope of this PR, but any reason the gcc-8 compiler doesn't have gfortran?

Comment (kennyweiss) [reply scripts/docker/dockerfile_gcc-8:8]: Also, is there a reason we're using gfortran-7 with our gcc-8 compiler image?

File: scripts/docker/dockerfile_gcc-8 [hunk 3]
```diff
@@ -1,25 +1,38 @@
...
+RUN sudo ln -f -s python3 /usr/bin/python # Compatibility with legacy Python
 
-RUN git clone --recursive https://github.com/LLNL/axom.git
+# Avoid building CMake from scratch
+RUN wget https://cmake.org/files/v3.10/cmake-3.10.1-Linux-x86_64.sh
+RUN sudo sh cmake-3.10.1-Linux-x86_64.sh --prefix=/usr --exclude-subdir
```
Comment (kennyweiss) [inline scripts/docker/dockerfile_gcc-8:13]: Another question -- is there a specific reason to use `cmake-3.10` ? Or is it just that we need a `CMake`, and this version is known to work?

File: scripts/docker/dockerfile_gcc-8 [hunk 4]
```diff
@@ -1,25 +1,38 @@
...
-WORKDIR "/home/axom/RAJA/build"
-RUN cmake ../. -DENABLE_TESTS=OFF
-RUN make 
-RUN sudo make install
+WORKDIR "/home/axom"
+USER axom
 
-WORKDIR "/home/axom/axom"
-RUN ./scripts/uberenv/uberenv.py --spack-config-dir=./scripts/uberenv/spack_configs/docker --spec=%gcc@8.1.0 --prefix=/home/axom/axom_tpls -k
+RUN git clone --recursive --branch $branch --single-branch --depth 1 https://github.com/LLNL/axom.git
```
Comment (kennyweiss) [inline scripts/docker/dockerfile_gcc-8:24]: I'm planning to look into this, but I'm assuming `$branch` is a user-settable variable within our new github action ?

File: scripts/docker/dockerfile_gcc-8 [hunk 5]
```diff
@@ -1,25 +1,38 @@
...
 
-WORKDIR "/home/axom"
+# New containers will use Python3
+RUN python3 ./axom/scripts/uberenv/uberenv.py --spack-config-dir=./axom/scripts/uberenv/spack_configs/docker/ubuntu16/ --project-json=./axom/scripts/uberenv/project.json \
+                                               --spec=%gcc@8.1.0 --prefix=/home/axom/serac_tpls -k
```
Comment (kennyweiss) [inline scripts/docker/dockerfile_gcc-8:28]: `serac_tpls` -> `axom_tpls` ?

File: scripts/docker/dockerfile_gcc-8 [hunk 6]
```diff
@@ -1,25 +1,38 @@
...
+
+# Make sure the new hostconfig worked
+RUN cd axom && python3 config-build.py -hc *.cmake -bp build && cd build && make -j8 VERBOSE=1 && make CTEST_OUTPUT_ON_FAILURE=1 test ARGS="-T Test -VV -j8"
```
Comment (kennyweiss) [inline scripts/docker/dockerfile_gcc-8:31]: Minor: What is the name of the `hc` file?  
I think uberenv prints it, right?

File: .github/workflows/docker_build_tpls.yml [hunk 7]
```diff
@@ -1,30 +1,131 @@
...
+        docker create --name extract_hc ${{ steps.repo_name.outputs.repo_plus_tag }}
+        mkdir -p extracted_hc
+        docker cp extract_hc:${{ env.HOSTCONFIG_LOC }} ./extracted_hc
+        docker rm extract_hc
+    
+    - name: Upload hostconfig
+      uses: actions/upload-artifact@v2
+      with:
+        name: ${{ env.TAG }}_hostconfigs
+        path: ./extracted_hc/export_hostconfig/*
```
Comment (kennyweiss) [inline .github/workflows/docker_build_tpls.yml:131]: Is the plan for us to manually download these artifact host-configs and create a branch / PR to update our CI docker images ?

**Review by kennyweiss — APPROVED** [review]
Thanks @hennis1  and @white238 -- this looks great!

File: scripts/docker/dockerfile_gcc-8 (hunk 4 above)
Comment (kennyweiss) [reply scripts/docker/dockerfile_gcc-8:24]: I found it -- 
image

File: .github/workflows/docker_build_tpls.yml [hunk 8]
```diff
@@ -1,30 +1,131 @@
...
+      run: echo "##[set-output name=branch;]$(echo ${GITHUB_REF#refs/heads/})"
+      id: extract_branch
+    - name: Get dockerfile name
+      shell: bash
+      run: echo "##[set-output name=filename;]$(echo ${DOCKERFILE_PREFIX}${TAG})"
+      id: dockerfile_name
+    - name: Get dockerhub repo name
+      shell: bash
+      run: |
+        echo ${REPO}:${TAG}_`date "+%m-%d-%y_%Hh-%Mm"`
```
Comment (kennyweiss) [inline .github/workflows/docker_build_tpls.yml:29]: Note: the time stamp appears to be UTC:
I.e. a job from a few hours ago ( 12/1/2020 ~ 4 PM PT) labels its TPL as:
```
Status: Downloaded newer image for axom/tpls:clang-10_12-02-20_00h-05m
```

File: .github/workflows/docker_build_tpls.yml (hunk 1 above)
Comment (white238) [reply .github/workflows/docker_build_tpls.yml]: Turns out conduit needed to be locked down and wasn't in the ubuntu-18 packages.yaml

File: scripts/docker/dockerfile_gcc-8 (hunk 2 above)
Comment (white238) [reply scripts/docker/dockerfile_gcc-8:8]: not sure if this was copied from somewhere else.  Now that it actually is in a good state we can look into things like this.

File: scripts/docker/dockerfile_gcc-8 (hunk 3 above)
Comment (white238) [reply scripts/docker/dockerfile_gcc-8:13]: known to work, copied from somewhere, not necessary, and should be updated.

File: scripts/docker/dockerfile_gcc-8 (hunk 6 above)
Comment (white238) [reply scripts/docker/dockerfile_gcc-8:31]: The host config is some oddly named file. For example, when I get the clang@10 host config from the workflow job  (it's an artifact), it is called `buildkitsandbox-linux-ubuntu18.04-x86_64-clang@10.0.0.cmake`.  I am not sure how stable that name is, specifically the "buildkitsandbox" part.

File: .github/workflows/docker_build_tpls.yml (hunk 7 above)
Comment (white238) [reply .github/workflows/docker_build_tpls.yml:131]: Yes, I asked @hennis1  to do it in this branch.  After he does that I can update the branch permissions when we are happy.

File: .github/workflows/docker_build_tpls.yml (hunk 8 above)
Comment (white238) [reply .github/workflows/docker_build_tpls.yml:29]: We are using it mainly as a unique identifier.  Do you want it more accurate in the PST sense?

Comment (kennyweiss) [reply .github/workflows/docker_build_tpls.yml:29]: No, I think UTC is probably the right choice.  I was just pointing it out.

File: scripts/docker/dockerfile_gcc-8 (hunk 6 above)
Comment (joshessman-llnl) [reply scripts/docker/dockerfile_gcc-8:31]: It should be pretty stable - the Docker build action that's used here is versioned and will always use buildx/buildkit.  The wildcard was added for the "manual" docker builds, in which the hostconfig would be prefixed with the (unstable) hash of the Docker image.

File: host-configs/docker/docker-linux-ubuntu16.04-ivybridge-gcc@8.1.0.cmake [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+set(CMAKE_CXX_FLAGS "-pthread" CACHE PATH "")
+
+set(BLT_CXX_STD "c++14" CACHE PATH "")
+
+#------------------------------------------------------------------------------
+# TPLs
+#------------------------------------------------------------------------------
+
+# Root directory for generated TPLs
+set(TPL_ROOT "/home/axom/serac_tpls/gcc-8.1.0" CACHE PATH "")
```
Comment (white238) [inline host-configs/docker/docker-linux-ubuntu16.04-ivybridge-gcc@8.1.0.cmake:33]: The new host-config here:

https://github.com/LLNL/axom/suites/1606131918/artifacts/29349708

doesn't have serac_tpls in the path.  It should fix the CMake error in the gcc@8 checks.

File: scripts/docker/dockerfile_gcc-8 (hunk 2 above)
Comment (white238) [reply scripts/docker/dockerfile_gcc-8:8]: #403

File: scripts/docker/dockerfile_gcc-8 (hunk 3 above)
Comment (white238) [reply scripts/docker/dockerfile_gcc-8:13]: #404

**Review by white238 — APPROVED** [review]

**Pull request merged**